bin/generate-gitlab-ci-yml.sh

```diff
@@ -0,0 +1,80 @@
+#! /usr/bin/env bash
+
+original_directory=$(pwd)
+. "${original_directory}/share/spack/setup-env.sh"
+
+if [ -z "${DOWNSTREAM_CI_REPO}" ] ; then
+    echo "ERROR: missing variable: DOWNSTREAM_CI_REPO" >&2
+    exit 1
+fi
+
+if [ -z "${SPACK_RELEASE_ENVIRONMENT_PATH}" ] ; then
+    echo "ERROR: missing variable: SPACK_RELEASE_ENVIRONMENT_PATH" >&2
+    exit 1
+fi
+
+if [ -z "${CDASH_AUTH_TOKEN}" ] ; then
+    echo "ERROR: missing variable: CDASH_AUTH_TOKEN" >&2
+    exit 1
+fi
+
+if [ -z "${SPACK_RELEASE_ENVIRONMENT_REPO}" ] ; then
+    echo "Assuming spack repo contains environment" >&2
+    env_repo_dir=${original_directory}
+else
+    echo "Cloning environment repo: ${SPACK_RELEASE_ENVIRONMENT_REPO}" >&2
+    mkdir /work
+    cd /work
+    git clone ${SPACK_RELEASE_ENVIRONMENT_REPO} envrepo
+    cd envrepo
+    env_repo_dir=$(pwd)
+fi
+
+current_branch="$CI_COMMIT_REF_NAME"
+
+# Because want to see generated gitlab-ci file as an artifact,
+# we need to write it within the spack repo cloned by gitlab-ci.
+gen_ci_dir="${original_directory}/ci-generation"
+gen_ci_file="${gen_ci_dir}/.gitlab-ci.yml"
+mkdir -p ${gen_ci_dir}
+
+env_dir="${env_repo_dir}/${SPACK_RELEASE_ENVIRONMENT_PATH}"
+
+if [ ! -f "${env_dir}/spack.yaml" ] ; then
+    echo "ERROR: Cannot find spack environment file in ${env_dir}"
+    exit 1
+fi
+
+cd $env_dir
+
+token_file=/tmp/cdash_auth_token
+echo ${CDASH_AUTH_TOKEN} > ${token_file}
+
+# This commands generates the .gitlab-ci.yml and creates buildgroup in cdash
+spack release-jobs --force --output-file ${gen_ci_file} --cdash-credentials ${token_file}
+
+cp ${gen_ci_file} "${original_directory}/.gitlab-ci.yml"
+
+git config --global user.email "robot@spack.io"
+git config --global user.name "Build Robot"
+
+commit_msg="Auto-generated commit testing ${current_branch} (${CI_COMMIT_SHA})"
+
+cd ${original_directory}
+echo "git status"
+git status
+echo "git branch"
+git branch -D ___multi_ci___ 2> /dev/null || true
+echo "git checkout"
+git checkout -b ___multi_ci___
+echo "git add"
+git add .gitlab-ci.yml
+echo "git commit"
+git commit -m "$commit_msg"
+echo "git commit-tree/reset"
+git reset "$( git commit-tree HEAD^{tree} -m ${commit_msg} )"
+echo "git status"
+git status
+echo "git push"
+git push --force "$DOWNSTREAM_CI_REPO" \
+    "___multi_ci___:$current_branch"
```

Comment:
1. Might want to add a prefix to the pushed branch name, maybe something like `multi-ci-$current_branch`.  Should make these generated branches easier to identify.

 1. Not sure if you've already thought of this, but at some point at the end of the generated workflow, we're going to want to come back and clean out these auto-generated branches.

Reply:
> 1. Might want to add a prefix to the pushed branch name, maybe something like `multi-ci-$current_branch`.  Should make these generated branches easier to identify.

Yeah, I think that could make sense.  Are these branches going to be publicly visible ultimately?  If so, it would be nice for people to recognize their branch names, in which case maybe I should tack something onto the end of the name rather than the beginning.  If not, it may not matter much.  But I agree that having something on there that helps us identify the branches created/pushed by the workflow makes sense.

> 2. Not sure if you've already thought of this, but at some point at the end of the generated workflow, we're going to want to come back and clean out these auto-generated branches.

It occurred to me at some point, but I haven't done anything about it yet.  Perhaps we could talk about it if/when we have the discussion I mentioned above.  I've been thinking it could be nice to have a history of the generated `.gitlab-ci.yml` files maintained somehow.  But I'm not married to the idea in any way.

Reply:
Regarding point 1: My thinking falls on two points: First, pushing a branch with nothing but a hash for the name does nothing to communicate the branch's purpose.  I don't expect human eyes to be looking at these branches regularly, but neither should someone poking around in these internals be scratching their heads wondering to themselves what these branches are all about?

Second, while we expect that in most use cases the downstream repo would exist exclusively for triggering the next phase of CI, I don't think we should ignore the possibility that someone would use a downstream repo that has other purposes.  In this case, we would certainly expect more human eyes to be on these generated branches, but more importantly, I'd want to take every measure to avoid name collisions.  Let's not assume we're the only automated process that is generating branches with hashes in their names.

Reply:
> Regarding point 1: My thinking falls on two points: First, pushing a branch with nothing but a hash for the name does nothing to communicate the branch's purpose. I don't expect human eyes to be looking at these branches regularly, but neither should someone poking around in these internals be scratching their heads wondering to themselves what these branches are all about?

Just to be clear, we're not pushing a branch with nothing but a hash for the name, right?  We're pushing a branch with whatever name the developer gave their own branch.  Now, maybe that's not the best idea, because it's not really their branch, but their branch with the appropriate `.gitlab-ci.yml` added. 

> Second, while we expect that in most use cases the downstream repo would exist exclusively for triggering the next phase of CI, I don't think we should ignore the possibility that someone would use a downstream repo that has other purposes. In this case, we would certainly expect more human eyes to be on these generated branches, but more importantly, I'd want to take every measure to avoid name collisions. Let's not assume we're the only automated process that is generating branches with hashes in their names.

Agreed, I'll add the prefix you suggested.